runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/LockFreeHistogramTest.java

```diff
@@ -0,0 +1,182 @@
+    histogram.update(5, 15, 25);
+    Optional<LockFreeHistogram.Snapshot> snapshot = histogram.getSnapshotAndReset();
+
+    LockFreeHistogram.OutlierStatistic expectedOverflow = LockFreeHistogram.OutlierStatistic.EMPTY;
+    LockFreeHistogram.OutlierStatistic expectedUnderflow = LockFreeHistogram.OutlierStatistic.EMPTY;
+    ImmutableLongArray expectedBuckets = ImmutableLongArray.of(1L, 1L, 1L);
+    LockFreeHistogram.Snapshot expectedSnapshot =
+        LockFreeHistogram.Snapshot.create(
+            expectedUnderflow, expectedOverflow, expectedBuckets, bucketType);
+
+    assertThat(snapshot.isPresent(), equalTo(true));
+    assertThat(snapshot.get(), equalTo(expectedSnapshot));
+  }
+
+  @Test
+  public void testUpdate_EmptySnapshot() {
+    HistogramData.BucketType bucketType = HistogramData.LinearBuckets.of(0, 10, 3);
+    LockFreeHistogram histogram =
+        new LockFreeHistogram(KV.of(MetricName.named("name", "namespace"), bucketType));
+    histogram.update(5, 15, 25);
+    Optional<LockFreeHistogram.Snapshot> snapshot_1 = histogram.getSnapshotAndReset();
+
+    assertThat(snapshot_1.isPresent(), equalTo(true));
+
+    Optional<LockFreeHistogram.Snapshot> snapshot_2 = histogram.getSnapshotAndReset();
+    assertThat(snapshot_2.isPresent(), equalTo(false));
+  }
+
+  /** A runnable records 200 values and then calls getSnapshotAndReset. */
+  private static class UpdateHistogramRunnable implements Runnable {
+    private final LockFreeHistogram histogram;
+    private final int val;
+    private Optional<LockFreeHistogram.Snapshot> snapshot;
+
+    private static final long valuesRecorded = 200L;
+
+    public UpdateHistogramRunnable(LockFreeHistogram histogram, int val) {
+      this.histogram = histogram;
+      this.val = val;
+      this.snapshot = Optional.empty();
+    }
+
+    @Override
+    public void run() {
+      for (long j = 0; j < valuesRecorded; j++) {
+        histogram.update(val);
+      }
+      snapshot = histogram.getSnapshotAndReset();
+    }
+
+    public long totalCountInSnapshot() {
+      if (snapshot.isPresent()) {
+        return snapshot.get().totalCount();
+      }
+      return 0;
+    }
+
+    public static long numValuesRecorded() {
+      return valuesRecorded;
+    }
+  }
+
+  @Test
+  public void testUpdateAndSnapshots_MultipleThreads() {
+    int numRunnables = 200;
+    ExecutorService executor = Executors.newFixedThreadPool(numRunnables);
+    HistogramData.BucketType bucketType = HistogramData.ExponentialBuckets.of(1, 10);
+    LockFreeHistogram histogram =
+        new LockFreeHistogram(KV.of(MetricName.named("name", "namespace"), bucketType));
+
+    UpdateHistogramRunnable[] runnables = new UpdateHistogramRunnable[numRunnables];
+    for (int i = 0; i < numRunnables; i++) {
+      runnables[i] = new UpdateHistogramRunnable(histogram, i);
+      executor.execute(runnables[i]);
+    }
+
+    try {
+      executor.shutdown();
+      executor.awaitTermination(30, TimeUnit.SECONDS);
```

Comment:
right now it is executing the tasks 1 by 1 not (possibly) concurrently since execute is being called in the loop

Consider using https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CountDownLatch.html

you can do something like the following to make the test more deterministic (not bound bu
`CountDownLatch testLatch = new CountDownLatch(howManyTimesYouWantTheRunnableToRun)`
pass `testLatch` into `UpdateHistogramRunnable` constructor
at the end of `UpdateHistogramRunnable#run()` method call `testsLatch.countDown()`
then call `testLatch.await()` (wait until this has ran `howManyTimesYouWantTheRunnableToRun`)
and then use ExecutorService#submit to submit each runnable, this does not block until the runnable is done

another way is to not call execute in the `loop`, build the list of `UpdateHistogramRunnable` then use `ExecutorService.invokeAll(UpdateHistogramRunnable)`.  I think right now it is executing 1 by 1.
based on lang specification invokeAll blocks but i haven't used it

Reply:
Done.